[REEF-1732] Build Metrics System

mandyshieh opened the pull request:
Enhances the capabilities of MetricService by adding the following
functionality:
 - Lets the user specify the type of the metric, as opposed to being restricted to Counters.
 - Lets the user maintain a timeseries and keep track of the history of
 updates for a given metric.
 - Added example implementations and functional tests to demonstrate
 usage.

JIRA:
 [REEF-1732](https://issues.apache.org/jira/browse/REEF-1732)

singlis commented on lang/cs/Org.Apache.REEF.Common/Org.Apache.REEF.Common.csproj:
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Hi,
We are using the *.DotNet projects, so these changes need to go into Apache.REEF.Common.DotNet.csproj. Sorry for the confusion. I have a PR out that deletes the old projects but it hasn't gone through yet.

singlis replied:
Actually if you only added cs files to the project, then no changes should be needed since the DotNet.csproj should automatically pickup the new files.

mandyshieh replied:
Double checked and everything seems to be in place in the Reef.DotNet.sln too 👍

singlis commented on lang/cs/Org.Apache.REEF.Common.Tests/Telemetry/MetricTests.cs:
```diff
@@ -0,0 +1,106 @@
+        {
+            var metrics = CreateMetrics();
+            metrics.TryRegisterMetric(new IntegerGauge("int1", "metric of type int", DateTime.Now.Ticks, 0));
+            Assert.Throws<ApplicationException>(() => metrics.Update(new DoubleGauge("int1", "new description", DateTime.Now.Ticks, 3)));
```
Do we want to add a new object every time an update occurs? I think the code should register a metric, get an instance to that metric and then set the value of that metric. Allocating a new object every time can be a large cost -- especially if doing so in a loop, etc.

mandyshieh replied:
In the case where we want to keep a history of the metric, a new object would have to be created at each update anyway, so this would save cost only in the case where we are not interested in keeping a record.. adding this comment to narrow the scope.

tyclintw replied:
@singlis has a point in regards to recreating the object to pass around the "name" and "description" which is redundant. If we could just pass around the next "T value" that would be ideal.

mandyshieh replied:
Changed so MetricData tracks updates directly made on the metric object. MetricData will create a MetricRecord which holds the value and timestamp to track updates, instead of copying the whole metric each time.

sharathmalladi commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,240 @@
+        {
+            lock (_metricLock)
+            {
+                foreach (var metric in metrics.GetMetrics())
```
Could we avoid the use of "var" when the data type is obvious? The reason is that expanding on the data type increases readability.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/DriverSystemState.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace Org.Apache.REEF.Common.Telemetry
+{
+    class DriverSystemState : MetricBase<string>
```
Previously, DriverMetrics was a class that only contained a string, so this class is a result of expanding on that to make it a metric of type string, though it would make sense to further expand and let the user decide what metrics the driver should collect. Will change string metric to a more generic name and expand IDriverMetrics.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/IntegerGauge.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace Org.Apache.REEF.Common.Telemetry
+{
+    class IntegerGauge : MetricBase<int>
```
I added some more primitive type metrics. Now there's counter, int, double, float, long, bool, and string available to use, and the user can implement more as needed.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/Counter.cs:
```diff
@@ -12,122 +16,54 @@
-    [Unstable("0.16", "This is a simple counter for evaluator metrics.")]
-    [DataContract]
-    internal sealed class Counter : ICounter
+    internal sealed class Counter : MetricBase<int>, ICounter
```
Renamed all to *Metric.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/ICounter.cs:
```diff
@@ -11,36 +15,18 @@
 
         /// <summary>
-        /// Increase the current counter value with the number specified.
+        /// Decresae the current counter value with the number specified.
```
fixed.

tyclintw commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,196 @@
+                Timestamp = timestamp;
+            }
+
+            public MetricRecord(MetricTracker metricData)
```
Just pass the IMetric, since the MetricData isn't referenced anywhere.

tyclintw commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,196 @@
+            UpdateMetric(metric);
+        }
+
+        public void OnError(Exception error)
```
At least log the error.

tyclintw commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,196 @@
+            return _records;
+        }
+
+        public void Subscribe(IMetric provider)
```
For all subscriptions there should be a corresponding Unsubscribe.

tyclintw commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,196 @@
+            UpdateRecords();
+        }
+
+        private void UpdateRecords()
```
If possible, try to stay away from internal states and take a functional programming approach. If you isolate all your logic to the function you have less of a chance for unwanted side-effects.
```csharp
private void UpdateRecords(IMetric metric)
{
    var newRecord = CreateMetricRecord(metric);
```

tyclintw commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,190 @@
+
+        public string Serialize()
+        {
+            lock (_metricLock)
```
The _metricLock usage is inconsistent. We'll want to synchronize a lot of this logic. I'll take with you offline.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Org.Apache.REEF.Common.csproj:
```diff
@@ -247,25 +247,32 @@ under the License.
-    <Compile Include="Telemetry\CountersData.cs" />
-    <Compile Include="Telemetry\Counters.cs" />
-    <Compile Include="Telemetry\CounterSinkThreshold.cs" />
+    <Compile Include="Telemetry\BooleanMetric.cs" />
```
Good point. I took out everything but counter because CounterMetric holds implementation for Increment and Decrement.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common.Tests/Telemetry/MetricTests.cs:
```diff
@@ -0,0 +1,128 @@
+        public void TestDuplicatedNames()
+        {
+            var metrics = CreateMetrics();
+            metrics.TryRegisterMetric(new CounterMetric("metric1", "metric description"));
```
fixed.

mandyshieh commented:
Hi Scott,
@tyclintw helped me come up with the current updates that eliminates the need for any locks. Please review and let us know what you think.

mandyshieh commented:
Hello,

I spent the day looking into EventCounters, and wanted to add a comment to share my findings. While I could get it to collect some data, I don't think it is completely in line with what we set out to achieve when we started building a metrics system for Reef:

- First and foremost, EventSource/EvenCounters seem to be designed for the user to define the ETW provider, so only for Windows applications. I couldn't find any examples of how event capturing was used in different OS's. 
- The above comment mentioned the viewing tool kept history; in the examples I found PerfView to be the primary tool used to view all the collected events over a timespan (is this what you meant?) The tool collects events on a specified interval, and calculates statistics on the execution times of events captured during that interval (mean, standard deviation, etc.) However, if I wanted to take these values and send them somewhere else, for example AML training service, I haven't found a way to do it easily.
- EventSource/EventCounter seem pretty confined to numbers and strings, I'm not sure there is an easy way to collect metrics of user defined types.

There were two original goals of this PR, one was to maintain a timeseries of updates of a single metrics, and the other was to support any type of metric that the user would specify. For the first one, I feel a push model is more suitable, and with help and many iterations it is finally not needing additional locks that might be a hit to performance while still preserving the entire history of updates.

Finally, to the last point, I feel this would probably be a sink implementation, where the metrics collected by MetricsService would be sent somewhere that is queryable by the user. Let me know what you think.

singlis commented:
Thanks Mandy for the information!

So EventCounters are cross plat, they are recommended to use over the Performance counters because of that reason. They do have roots from Performance Counters, which can be a bit confusing.  Please refer to the reply here by vancem:
https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Tracing/documentation/EventCounterTutorial.md

As for PerfView, yes that is my understanding for the current toolset. For Windows, you maybe able to also use the Performance Monitor, but I haven't been able to confirm that. There should be an .net API that allows for collection of these values, please see the EventListener:
https://msdn.microsoft.com/en-us/library/system.diagnostics.tracing.eventlistener(v=vs.110).aspx
It handles a OnEventWritten function when an event is written. 

Also EventSource looks to support different types of formats looking at the EventFieldFormat enum:
https://msdn.microsoft.com/en-us/library/system.diagnostics.tracing.eventfieldformat(v=vs.110).aspx

Ok, so now looking more at this. It looks like if you use EventCounters, they will notify you when an event counter updates via the EventListener. So this is a push model. You can collect this data like you are doing now on the evaluator and submit that via the heartbeat. The driver would then aggregate this data.

Given this information, would this work? Sounds like it matches the requirements in that you would not miss a data point as you could via the pull method. Also what I like about the EventCounters is that its utilizing the .net core api. 

Let me know -- Im happy to also review what you have.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+        internal MetricTracker UpdateMetric(MetricTracker metric)
+        {
+            _mirror = metric.GetMetric();
+            if (metric.ChangesSinceLastSink > 0)
```
This doesnt look to be thread-safe - if two threads came here and checked if metric.ChangesSinceLastSink > 0, they would both hit the last code of incrementing ChangesSinceLastSink. So you would miss an event if history were being tracked.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+        private void UpdateRecords()
+        {
+            var newRecord = CreateMetricRecord(_mirror);
+            if (_keepUpdateHistory)
```
_keepUpdateHistory is set with the metric tracker. Does this mean we are allocating a ConcurrentQueue each time an update comes through?

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            _mirror = metric;
```
Subscribe should handle assigning _mirror.

mandyshieh replied:
fixed.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+            ChangesSinceLastSink = 0;
+            _keepUpdateHistory = metric.IsImmutable;
+            _records = new ConcurrentQueue<MetricRecord>();
+            _records.Enqueue(CreateMetricRecord(metric));
```
Is this Create just to create a snapshot of the initial state?

mandyshieh replied:
Yup, keeping a record of the first value.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+        }
+
+        [JsonConstructor]
+        internal MetricTracker(IMetric _mirRor, int changesSinceLastSink, ConcurrentQueue<MetricRecord> _recorDs)
```
I would change all incoming parameters to not have and underscore. Underscore is usually reserved notation for class members. I would also remove the casing for _mirRor and _recorDs.

mandyshieh replied:
So this was done because JsonConstructor needs the parameters to match the fields. For the sake of neatness, I've changed all the fields to internal accessibility in Pascal case so that the parameters can all be camel case.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+        internal MetricTracker(IMetric metric)
+        {
+            Subscribe(metric);
+            _mirror = metric;
```
minor nit - I would rather see _mirror just called _metric, I think that may make more sense as you look through the code.

mandyshieh replied:
fixed.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+            }
+            else
+            {
+                _records = new ConcurrentQueue<MetricRecord>();
```
Talked about this offline. 1) We are allocating a concurrent queue each time we increment. 2) This assignment and enqueueing are not thread safe and could cause some undersirable outcomes. Mandy was looking into creating two different types of trackers to handle this.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+        internal void UpdateMetric(IMetric me)
+        {
+            ChangesSinceLastSink++;
+            _mirror = me;
```
You are right, there is only one instance of the metric object so it's not needed. Took it out here and other places.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,221 @@
+        }
+
+        [JsonObject]
+        public struct MetricRecord
```
Made it a class to avoid copies during flush dequeue.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/CounterMetric.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            _typedValue += number;
+            _timestamp = DateTime.Now.Ticks;
+            _tracker.OnNext(this);
```
Also if you change the code to capture with local vars, be careful with things like:
_typedValue +=number
var newValue = _typedValue
...
As this too is also not thread safe as a thread could have changed the _typeValue by the time you capture it to a local var newValue. You really want things to happen in one line. Doing an Interlocked.Add will at least return the new value. So the above code will become:
var newValue = Interlocked.Add(ref _typedValue, value);

mandyshieh commented:
Here are a list of areas that are identified as potentially "thread-dangerous" and how it has been addressed:

**Updating metrics and creating records:**
Interlocked is used in all cases of updating metric values. I added primitive children classes to MetricBase<T> for the pirmitive type Interlocked.Exchanges(int, double, long, float), and a MetricClass<T> for reference type Interlocked.Exchange(string and other classes). Only if all of these fail will a lock(_metricLock) be used.
The updated value is then passed to the tracker to create a new record, which also records a timestamp, instead of getting the timestamp from the metric itself.

**Passing metrics from the evaluator to the driver:**
 - KeepUpdateRecord == true: Each tracker maintains a ConcurrentQueue(q1) of records. On the heartbeat, each tracker create a new ConcurrentQueue(q2), moves the contents of q1 to q2 one by one until q1 is empty, then passes q2 to the driver. In the meantime, new records can still be added to q1 safely.
 - KeepUpdateRecord == false: On the Evaluator side, no records will be created and q1 will be empty until the heartbeat. On the heartbeat, one single record will be created from the value held by the metric object, the record will be put in q2, then passed.

**Updating metrics received from evaluator when multiple evaluators try to update the same metric:**
 - KeepUpdateRecord == true: records received from the evaluator are added to the driver's ConcurrentQueue one by one.
 - KeepUpdateRecord == false: Driver's records are reassigned to the received records using Interlocked.Exchange.

Sinking metrics on driver side:
 - same as Passing metrics from Evaluator to Driver

mandyshieh commented:
I realized that boxing and unboxing wouldn't work how I initially thought after our discussion because _typedValue variable doesn't reflect changes once it's boxed into an object, so I instead reverted back to individual primitive implementations, each with its own Interlocked.Exchange call, plus a MetricClass that takes reference type exchanges. It should be water-tight thread-safe now, let me know what you think :)

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,235 @@
+        internal ConcurrentQueue<MetricRecord> FlushChangesSinceLastSink()
+        {
+            ConcurrentQueue<MetricRecord> records = new ConcurrentQueue<MetricRecord>();
+            if(!Records.IsEmpty)
```
Originally Records was a private field of the class, and the way to access it was through GetMetricRecords that would either return the private _records OR return a queue with a single record. In other words, when not keeping history, private _records will always be empty, but when someone asks for the metric records through the api, they will always get at least one value.
Then _records became internal Records to avoid messy casing in the json constructor, but if we are ok with naming private fields in Pascal case then I think logically it would be clearer.

mandyshieh replied:
Keeping the Records.IsEmpty check for now because if GetMetricRecords() were used, an additional ConcurrentQueue would be allocated when the Records field is empty.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,235 @@
+        {
+            Subscribe(metric);
+            ChangesSinceLastSink = 0;
+            KeepUpdateHistory = metric.IsImmutable;
```
fixed.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricBase.cs:
```diff
@@ -0,0 +1,114 @@
+        {
+            Name = name;
+            Description = description;
+            _typedValue = value;
```
At the point where driver reconstructs the Metric object from the json string using this json constructor, the isImmutable information won't be useful anymore, since the tracker already has all the information, but for the sake of consistency I've added it.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/LongMetric.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public override void AssignNewValue(object val)
+        {
+            if (val.GetType() != _typedValue.GetType())
```
done.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,235 @@
+        /// <param name="value">Value of the new record.</param>
+        public void Track(object value)
+        {
+            ChangesSinceLastSink++;
```
fixed.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,235 @@
+        /// If KeepUpdateHistory is true, it will return all the records; otherwise, it will returen one record with the most recent value.
+        /// </summary>
+        /// <returns>The history of the metric records.</returns>
+        internal ConcurrentQueue<MetricRecord> GetMetricRecords()
```
fixed. Noting that some casting is required in UpdateMetric(MetricTracker)

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,235 @@
+        /// Flush records currently held in the records queue.
+        /// </summary>
+        /// <returns>A queue containing all the flushed records.</returns>
+        internal ConcurrentQueue<MetricRecord> FlushChangesSinceLastSink()
```
I did IEnumerable instead to enable linq operations.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,235 @@
+                // Records will be empty only on eval side when tracker doesn't keep history.
+                records.Enqueue(CreateMetricRecord(Metric));
+            }
+            ChangesSinceLastSink = 0;
```
Fixed except for in the ctor initialization.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,235 @@
+        /// If KeepUpdateHistory is true, it will return all the records; otherwise, it will returen one record with the most recent value.
+        /// </summary>
+        /// <returns>The history of the metric records.</returns>
+        internal ConcurrentQueue<MetricRecord> GetMetricRecords()
```
done.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,168 @@
+        /// Called when Evaluator is sending metrics information to Driver.
+        /// </summary>
+        /// <returns>Queue of trackers containing metric records.</returns>
+        public ConcurrentQueue<MetricTracker> FlushMetricTrackers()
```
IEnumerable to match Serialize param type where it's used.

mandyshieh commented on lang/cs/Org.Apache.REEF.IO.Tests/Org.Apache.REEF.IO.Tests.csproj:
```diff
@@ -33,5 +33,9 @@ under the License.
     <ProjectReference Include="..\Org.Apache.REEF.Utilities\Org.Apache.REEF.Utilities.csproj" />
```
Actually I can only run the tests with the update.. it seems it's only me though. Reverted as well in REEF.Tests.csproj.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricBase.cs:
```diff
@@ -0,0 +1,168 @@
+
+        public string Name
+        {
+            get; internal set;
```
these should be private set right? Ideally you want only the constructor to set these and not allow anyone to modify.

mandyshieh replied:
This is used set properties when IEvaluatorMetrics calls CreateAndRegisterMetric<T> where the metric's parameterless constructor is called.

singlis replied:
OK - see comment above in the CreateAndRegisterMetric function. I think if you change that to calling the constructor instead of setting the properties it will work.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricBase.cs:
```diff
@@ -0,0 +1,168 @@
+    /// <typeparam name="T"></typeparam>
+    public class MetricBase<T> : MetricBase
+    {
+        protected T _typedValue;
```
You should be able to remove this and continue using the untyped value in MetricBase to store the value. Then Value will cast to type T.

mandyshieh replied:
_typedValue was made mainly so that Interlocked methods would call the specific types. Most would still work with _value since there's the Exchange(ref object, object), but the Add(ref int, int) for Increment/Decrement in CounterMetric would fail.

singlis replied:
I know we talked more about this offline, is _typedValue going away?

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/DriverMetrics.cs:
```diff
@@ -15,25 +15,47 @@
+        public T CreateAndRegisterMetric<T>(string name, string description, bool keepUpdateHistory) 
+            where T : MetricBase, new()
+        {
+            var metric = new T
```
Minor nit - You can make this be T(name, description, keepUpdateHistory). Then you can have private setters on MetricBase.

mandyshieh replied:
So this is a restriction on C# generics that has been discussed here:
https://stackoverflow.com/questions/1682310/c-sharp-generics-problem-newing-up-the-generic-type-with-parameters-in-the-con

If the constructor were to be called directly, it'd need to be passed  as a parameter, like so:
        public T CreateAndRegisterMetric<T>(Func<string, string, bool, T> del, string name, string description, bool keepUpdateHistory) where T : MetricBase
        {
            var me = del(name, description, keepUpdateHistory);
            return me;
        }
which I don't think is a good idea since there's no way to ensure the function that gets passed in is actually the constructor that we want-- really method that takes three arguments and returns T would fly.
---
Another option is 

var metric = Activator.CreateInstance(typeof(T), name, description, keepUpdateHistory) as T

but here we'd trust that the derived class has a constructor that takes three arguments of the correct types. It's safer than the first method because at least we'll know it's a constructor. If no constructors match, this will throw. For all of our primitive types this would work, but if someone were to derive from the MetricClass<T> type, they'd need to add their own constructor, otherwise their metric wouldn't be able to register.
---
All this is to say, internal setters seemed to me like a reasonable compromise to avoid user error when creating and registering metrics. Comments and criticism welcome :P

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/IMetric.cs:
```diff
@@ -0,0 +1,62 @@
+        /// </summary>
+        /// <param name="val">Value to assign to the metric.</param>
+        /// <returns></returns>
+        void AssignNewValue(object val);
```
Is this needed at the interface level?

mandyshieh replied:
removed, and made some changes to DriverMetrics to be consistent with this change.

singlis commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsService.cs:
```diff
@@ -28,75 +28,73 @@
                 try
                 {
-                    Task.Run(() => s.Sink(metrics));
+                    //// Task.Run(() => s.Sink(metrics));
```
Should this be removed?

mandyshieh replied:
done.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/DoubleMetric.cs:
```diff
@@ -0,0 +1,51 @@
+        {
+        }
+
+        public override void AssignNewValue(object val)
```
good idea, done!

motus commented:
@singlis @mandyshieh Sure! I am looking at it now.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,238 @@
+        private IMetric Metric;
+
+        [JsonProperty]
+        internal bool KeepUpdateHistory;
```
can be `readonly`.

moreover - the only purpose of this variable seems to be to guard the `Records` queue. I think we can just allow `Records` to be null instead. then we'll have fewer member variables and our code simplifies a bit (i'll add some examples below)

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,238 @@
+        /// If KeepUpdateHistory is true, it will return all the records; otherwise, it will returen one record with the most recent value.
+        /// </summary>
+        /// <returns>The history of the metric records.</returns>
+        internal ICollection GetMetricRecords()
```
why not return `IEnumerable<MetricRecord>`?

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,238 @@
+            if (KeepUpdateHistory)
+            {
+                Records.Enqueue(CreateMetricRecord(value));
+            }
```
if we allow `Records` to be null, 4 lines above can be reduced to just one:
```C#
Records?.Enqueue(CreateMetricRecord(value));
```

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <summary>
+        /// Registration of metrics
+        /// </summary>
+        private ConcurrentDictionary<string, MetricTracker> _metricsMap = new ConcurrentDictionary<string, MetricTracker>();
```
`readonly`?

mandyshieh replied:
fixed.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,150 @@
+        public IEnumerable<KeyValuePair<string, MetricTracker.MetricRecord>> FlushMetricRecords()
+        {
+            // for each metric, flush the records and create key value pairs
+            return _metricsMap.SelectMany(kv => kv.Value.FlushChangesSinceLastSink().Select(r => new KeyValuePair<string, MetricTracker.MetricRecord>(kv.Key, r)));
```
here and everywhere: the line is too long. we try to keep your lines no longer than 120 characters; in most cases, it's below 100.

mandyshieh commented:
Hi @motus,
I've made the change so that MetricTracker's field Records will be null if the metric does not keep history, which simplifies the code somewhat. I kept the KeeupUpdateHistory field in the tracker because it is needed when the driver receives new metric data to decide whether to append or swap out the records.
Regarding ValueUntyped and non-generic IMetric, this is to enable all implementations of IMetric to be stored in the same collection. If we do IMetric<T>, then MetricTracker would also need to be templatized, and then MetricsData wouldn't be able to store all the trackers of different types in a single dict. To make up for it, EvaluatorMetrics and DriverMetrics enforces a typed implementation in CreateAndRegister. Hope this makes it more clear.

motus commented:
Hi @mandyshieh, I am still trying to understand the logic, but I have a feeling that we can be more type safe in many places. I've played with the code a bit, and pushed some of my changes to my private branch. Please check it out:

https://github.com/apache/reef/compare/d4d84afb5a86d8cb3df0770763a30a2398ea2f9a...motus:REEF-1732?expand=1

There are still a few things that I cannot figure out (e.g. how to get rid of the redundant type parameter in `IDriverMetrics.CreateAndRegisterMetric()`, so please take a look and let me know what you think!

Thank you!

mandyshieh commented:
The current iteration is actually the result of trying to solve for getting rid of the redundant generic param for CreateAndRegister<T,U>, as seen here: 
https://github.com/mandyshieh/reef/commit/f67b15abd1993ff217cec836cab0b3afb11c6956

@singlis and I had a lengthy discussion about this. Ideally, it'd be better if the values were stored as specific types, but that would require the user to specify that type along with the metric class on creation(`CreateAndRegisterMetric<IntegerMetric, int>() `instead of just `CreateAndRegisterMetric<IntegerMetric>()`). In the end we concluded that the lesser of two evils would be to make IMetric non-generic to eliminate the possibility of error in CreateAndRegister<T,U> when T and U don't match, especially since this PR contains concrete implementations for most cases so the user doesn't need to implement their own IMetric<U>.

@singlis feel free to chime in if there are points I didn't cover well. Thanks!

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,232 @@
+        /// <returns>The history of the metric records.</returns>
+        internal IEnumerable<MetricRecord> GetMetricRecords()
+        {
+            if (Records.IsEmpty)
```
shouldn't we check for `KeepUpdateHistory` instead? I would write something like
```C#
if (KeepUpdateHistory)
{
    return Records;
}
return new[] { CreateMetricRecord(Metric) };
```

mandyshieh replied:
The logic behind this is we only want to return CreateMetricRecord(Metric) when KeepUpdateHistory==false on the eval side. When eval passes metrics to the driver, there will always at least be one item in Records. The justification is if one were to call GetMetricRecords() from the driver, we want to return the last record that was passed from the evaluator, instead of the most current value of the metric itself.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,232 @@
+        [JsonObject]
+        public class MetricRecord
+        {
+            private object _value;
```
`readonly`?

mandyshieh replied:
fixed.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,232 @@
+                    return _value;
+                }
+            }
+            [JsonProperty]
```
add a blank line after a `}`

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,150 @@
+    /// </summary>
+    public sealed class MetricsData : IMetrics
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricsData));
```
logging is not used anywhere in this class

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,150 @@
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(MetricsData));
+
+        JsonSerializerSettings settings = new JsonSerializerSettings()
```
can be `private static readonly`

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,150 @@
+
+        public bool TryGetMetric(string name, out IMetric me)
+        {
+            if (!_metricsMap.TryGetValue(name, out MetricTracker tracker))
```
maybe just
```C#
bool success = _metricsMap.TryGetValue(name, out MetricTracker tracker);
metric = tracker?.GetMetric();
return success;
```
?

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricsData.cs:
```diff
@@ -0,0 +1,150 @@
+        /// <returns>Queue of trackers containing metric records.</returns>
+        internal IEnumerable<MetricTracker> FlushMetricTrackers()
+        {
+            return new ConcurrentQueue<MetricTracker>(_metricsMap.Select(kv => new MetricTracker(kv.Value.GetMetric(), kv.Value.ChangesSinceLastSink, (ConcurrentQueue<MetricTracker.MetricRecord>)kv.Value.FlushChangesSinceLastSink(), kv.Value.KeepUpdateHistory)));
```
long line! also, we can get rid of that cast if we make `MetricTracker` constructor accept `IEnumerable` as parameter

motus commented on lang/cs/Org.Apache.REEF.Tests/Functional/Telemetry/MetricsDriver.cs:
```diff
@@ -135,11 +139,12 @@ public void OnError(Exception error)
         {
-            var driverMetrics = new DriverMetrics(EventPrefix + systemState, DateTime.Now);
+            _driverMetrics.TryGetMetric(DriverMetrics.DriverStateMetric, out IMetric stateMetric);
+            ((StringMetric)stateMetric).AssignNewValue(EventPrefix + systemState.ToString());
```
this is where I would love to have some stronger typing! if we define our `.TryGetMetric()` method everywhere as
```C#
bool TryGetMetric<T>(string name, out T metric) where T : IMetric
```
then we can write just
```C#
_driverMetrics.TryGetMetric(DriverMetrics.DriverStateMetric, out StringMetric stateMetric);
stateMetric?.AssignNewValue(EventPrefix + systemState);
```
I am sure we can do that.
(also note that the explicit `.ToString()` was redundant)

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/CounterMetric.cs:
```diff
@@ -0,0 +1,60 @@
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
```
removed here and elsewhere.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/CounterMetric.cs:
```diff
@@ -0,0 +1,60 @@
+    /// </summary>
+    public sealed class CounterMetric : MetricBase<int>, ICounter
+    {
+        public CounterMetric() : base()
```
removed here and in the sibling classes

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/DoubleMetric.cs:
```diff
@@ -0,0 +1,50 @@
+        {
+        }
+
+        public override void AssignNewValue(double val)
```
fixed here and elsewhere.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/DriverMetrics.cs:
```diff
@@ -15,25 +15,47 @@
     public sealed class DriverMetrics : IDriverMetrics
     {
-        public DriverMetrics(string systemState, DateTime timeUpdated)
+        private MetricsData _metricsData;
```
fixed.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/IMetrics.cs:
```diff
@@ -0,0 +1,61 @@
+        /// Add a metric object to this collection.
+        /// </summary>
+        /// <param name="me">The metric object to add.</param>
+        void RegisterMetric(IMetric me);
```
fixed here and elsewhere

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/IDriverMetrics.cs:
```diff
@@ -21,18 +21,24 @@
 {
-    [Unstable("0.16", "This is to build a simple metrics with system state only. More metrics will be added in future.")]
     [DefaultImplementation(typeof(DriverMetrics))]
     public interface IDriverMetrics
```
Base interface IMetrics, and renamed interface for MetricsData to IMetricSet

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricBase.cs:
```diff
@@ -0,0 +1,135 @@
+        {
+        }
+
+        public MetricBase(string name, string description, bool keepUpdateHistory = true)
```
fixed

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricBase.cs:
```diff
@@ -0,0 +1,135 @@
+            KeepUpdateHistory = keepUpdateHistory;
+        }
+
+        public IDisposable Subscribe(ITracker observer)
```
fixed.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,232 @@
+                }
+                else
+                {
+                    Interlocked.Exchange(ref Records, (ConcurrentQueue<MetricRecord>)recordsToAdd);
```
fixed.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -0,0 +1,232 @@
+                    Interlocked.Exchange(ref Records, (ConcurrentQueue<MetricRecord>)recordsToAdd);
+                }
+            }
+            Interlocked.Add(ref ChangesSinceLastSink, metric.ChangesSinceLastSink);
```
I took out the ChangesSinceLastSink variable to avoid having to use a lock (@singlis and I discussed this at length and wanted to avoid the performance hit.) Instead the sink threshold directly checks the number of cached records to decide whether to sink or not.

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricBase.cs:
```diff
@@ -0,0 +1,135 @@
+        }
+
+        [JsonConstructor]
+        public MetricBase(string name, string description, T value, bool keepUpdateHistory)
```
The main purpose for reconstructing the Metric object on the driver side is for the metatdata, specifically the name, but on second thought it'd make more sense to pull the name out as a field of the tracker.
To summarize the fix, the metric object will not be reconstructed on the driver; all important values should be passed as record instead.

mandyshieh commented:
Hi @motus,
The changes you've requested have been pushed. In addition to cleaning up unused directives, long lines, etc., there are two main changes:
1. Took out the int ChangesSinceLastSink variable. Instead, the sink checks thresholds by directly counting from the cache of records. That way, there is no chance that a thread might interrupt during updates causing inconsistencies between records and changes.
2. The metric object itself is not reconstructed once records have been passed to the Driver Side, since it is never used.
Hope this addresses your concerns! Let me know what you think.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -46,12 +45,8 @@ public sealed class MetricTracker : ITracker
-        [JsonProperty]
-        internal int ChangesSinceLastSink;
+
+        private IMetric Metric;
```
naming conventions: use `_metric` for private members

mandyshieh replied:
fixed.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -26,17 +25,17 @@
 
         [JsonProperty]
-        private IMetric Metric;
+        public string MetricName;
```
`readonly`?

mandyshieh replied:
fixed.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -179,7 +172,6 @@ public void Unsubscribe()
-            Interlocked.Increment(ref ChangesSinceLastSink);
             Records?.Enqueue(CreateMetricRecord(value));
         }
 
```
`CreateMetricRecord()` can be `static`

mandyshieh replied:
makes sense, fixed.

motus commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -196,7 +188,7 @@ private MetricRecord CreateMetricRecord(object val)
         [JsonObject]
         public class MetricRecord
```
maybe just
```C#
[JsonObject]
public class MetricRecord
{
    [JsonProperty]
    public readonly object Value;

    [JsonProperty]
    public readonly long Timestamp;

    [JsonConstructor]
    public MetricRecord(object value, long? timestamp = null)
    {
        Value = value;
        Timestamp = timestamp.GetValueOrDefault(DateTime.Now.Ticks);
    }

    public MetricRecord(IMetric metric)
    {
        Timestamp = DateTime.Now.Ticks;
        Value = metric.ValueUntyped;
    }
}
```

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/IMetrics.cs:
```diff
@@ -5,57 +5,44 @@
 // "License"); you may not use this file except in compliance
 // with the License.  You may obtain a copy of the License at
-//
+// 
```
All cleaned up

mandyshieh commented on lang/cs/Org.Apache.REEF.Common/Telemetry/MetricTracker.cs:
```diff
@@ -26,17 +25,17 @@
     [JsonObject]
     public sealed class MetricTracker : ITracker
     {
         private static readonly Logger Logger = Logger.GetLogger(typeof(MetricTracker));
```
deleted along with using reference

mandyshieh commented:
All comments addressed and ready to go! :D
